internal/argocd/cluster/cluster.go

```diff
@@ -39,67 +44,154 @@ func (m *Manager) UpdateClusterConnectionInfo(agentName, status string, t time.T
+	if err := setClusterInfo(m.ctx, m.kubeclient, m.namespace, cluster.Server, m.redisAddress, agentName, cluster.Name, m.redisCompressionType,
 		&appv1.ClusterInfo{
 			ConnectionState: appv1.ConnectionState{
 				Status:     status,
 				Message:    fmt.Sprintf("Agent: '%s' is %s with principal", agentName, state),
-				ModifiedAt: &metav1.Time{Time: t}}},
-	); err != nil {
-		log().Errorf("Failed to update connection info in Cluster: '%s' mapped with Agent: '%s'. Error: %v", cluster.Name, agentName, err)
+				ModifiedAt: &metav1.Time{Time: t},
+			},
+		}); err != nil {
+		log().Errorf("failed to refresh connection info in cluster: '%s' mapped with agent: '%s'. Error: %v", cluster.Name, agentName, err)
 		return
 	}
 
 	log().Infof("Updated connection status to '%s' in Cluster: '%s' mapped with Agent: '%s'", status, cluster.Name, agentName)
 }
 
-// refreshClusterConnectionInfo gets latest cluster info from cache and re-saves it to avoid deletion of info
-// by ArgoCD after cache expiration time duration (i.e. 10 Minute)
-func (m *Manager) refreshClusterConnectionInfo() {
+// refreshClusterInfo gets latest cluster info from cache and re-saves it to avoid deletion of info
+// by Argo CD after cache expiration time duration (i.e. 10 minutes)
+func (m *Manager) refreshClusterInfo() {
 	m.mutex.Lock()
 	defer m.mutex.Unlock()
 
-	// iterate through all clusters
+	// Iterate through all clusters.
 	for agentName, cluster := range m.clusters {
-		// Create Redis client and cache
-		redisOptions := &redis.Options{Addr: m.redisAddress}
+		clusterInfo, err := GetClusterInfo(m.ctx, m.kubeclient, m.namespace, cluster.Server, m.redisAddress, m.redisCompressionType)
+		if err != nil {
+			if !errors.Is(err, cacheutil.ErrCacheMiss) {
+				log().Errorf("failed to get connection info from cluster: '%s' mapped with agent: '%s'. Error: %v", cluster.Name, agentName, err)
+			}
+			continue
+		}
 
-		if err := common.SetOptionalRedisPasswordFromKubeConfig(m.ctx, m.kubeclient, m.namespace, redisOptions); err != nil {
-			log().Errorf("Failed to fetch & set redis password for namespace %s: %v", m.namespace, err)
+		// Re-save same info.
+		if err := setClusterInfo(m.ctx, m.kubeclient, m.namespace, cluster.Server, m.redisAddress,
+			agentName, cluster.Name, m.redisCompressionType, clusterInfo); err != nil {
+			log().Errorf("failed to refresh connection info in cluster: '%s' mapped with agent: '%s'. Error: %v", cluster.Name, agentName, err)
+			continue
 		}
-		redisClient := redis.NewClient(redisOptions)
+	}
+}
 
-		cache := appstatecache.NewCache(cacheutil.NewCache(
-			cacheutil.NewRedisCache(redisClient, time.Minute, m.redisCompressionType)), time.Minute)
+// SetClusterCacheStats updates cluster cache info with Application, Resource and API counts in principal.
+// This is called when principal receives clusterCacheInfoUpdate event from agent.
+func (m *Manager) SetClusterCacheStats(clusterInfo appv1.ClusterInfo, agentName string) {
```

Comment:
Nit: we should return errors here, rather than just logging them. (If calling functions don't want to return them, that's up to them).

This will also allow the unit tests that call this function to verify no errors occurred.